src/pyggdrasil/tree_inference/_mcmc_sampler.py

```diff
@@ -0,0 +1,164 @@
+    data: MutationMatrix,
+    num_samples: int,
+    output_dir: Path,
+    num_burn_in: int = 0,
+    thinning: int = 0,
+    iteration: int = 0,
+    **kwargs,
+) -> None:
+    """Sample mutation trees according to the SCITE model.
+
+    Args:
+        rng_key: random key for the MCMC sampler
+        init_tree: initial tree to start the MCMC sampler from
+        error_rates: \theta = (\alpha, \beta) error rates
+        move_probs: probabilities for each move
+        data: observed mutation matrix to calculate the log-probability of,
+            given current tree
+        num_samples: number of samples to return
+        output_dir: directory to save samples to
+        num_burn_in: number of samples to discard before returning samples
+        thinning: number of samples to discard between samples
+        iteration: sample numer in chain, for restarting
+
+    Returns:
+        None
+    """
+
+    logging.info("Starting MCMC sampler.")
+
+    # curry logprobability function
+    logprobability_fn = logprob.create_logprob(data, error_rates)
+
+    # get initial state
+    init_state = (
+        iteration,
+        rng_key,
+        init_tree.tree_topology,
+        init_tree.labels,
+        logprobability_fn(init_tree),
+    )
+
+    logging.info(f"Initial state: {init_state}.")
+    logging.info("Starting MCMC loop.")
+
+    # define loop body
+    def body(
+        state: Tuple[int, JAXRandomKey, Array, Array, float],
+    ) -> Tuple[int, JAXRandomKey, Array, Array, float]:
+        """Body of the MCMC loop.
+
+        Args:
+            state: tuple containing the current state of the MCMC sampler
+                - iteration: current iteration
+                - rng_key: random key for the MCMC sampler
+                - Array: current tree, adjacency matrix
+                - Array: labels of the tree
+                - float: log-probability of the current tree
+
+        Returns:
+                 updated state: tuple containing the current state of the MCMC sampler
+                - iteration: current iteration
+                - rng_key: random key for the MCMC sampler
+                - Array: current tree, adjacency matrix
+                - Array: labels of the tree
+                - float: log-probability of the current tree
+        """
+
+        # get current state
+        iter_sample, rng_key_body, topo, labels, logprobability = state
+        iter_sample = iter_sample + 1
+
+        # make Tree
+        tree = Tree(tree_topology=topo, labels=labels)
+
+        # split random key to use one
+        rng_key_run, rng_key_sample = jax.random.split(rng_key_body)
+
+        # mcmc kernel
+        tree, logprobability = mcmc._mcmc_kernel(
+            rng_key_sample,
+            tree,
+            data,
+            error_rates,
+            move_probs,
+            logprobability_fn,
+            logprobability,
+        )
+        logging.info("Iteration: %d, log-probability: %f", iter_sample, logprobability)
```

Comment:
This is one way of doing it, another one is via format strings (and then you can specify e.g., precision in which `float` is reported.)

Reply:
Changed to 4 digits precision.